crates/pallets/starknet/src/tests/invoke_tx.rs

```diff
@@ -617,3 +620,285 @@ fn storage_changes_should_revert_on_transaction_revert() {
+
+        // Ethereum fee token contract address
+        let eth_fee_contract_address =
+            ContractAddress(PatriciaKey(StarkFelt::try_from(ETH_FEE_TOKEN_ADDRESS).unwrap()));
+
+        // Starknet fee token contract address
+        let strk_fee_contract_address =
+            ContractAddress(PatriciaKey(StarkFelt::try_from(STRK_FEE_TOKEN_ADDRESS).unwrap()));
+
+        let mut transaction = get_invoke_v3_dummy(Starknet::chain_id(), NONCE_ZERO);
+
+        if let starknet_api::transaction::InvokeTransaction::V3(tx) = &mut transaction.tx {
+            tx.sender_address = sender_address;
+        };
+
+        set_sender_erc20_balance_to_zero(sender_address, eth_fee_contract_address);
+        set_sender_erc20_balance_to_zero(sender_address, strk_fee_contract_address);
+
+        assert_err!(Starknet::invoke(none_origin, transaction), Error::<MockRuntime>::TransactionExecutionFailed);
+    });
+}
+
+#[test]
+fn given_hardcoded_contract_run_invoke_v3_without_eth_with_strk_fee_token_it_works() {
+    new_test_ext::<MockRuntime>().execute_with(|| {
+        basic_test_setup(2);
+
+        let none_origin = RuntimeOrigin::none();
+
+        // Account that gonna make the transaction
+        let sender_address = get_account_address(None, AccountType::V0(AccountTypeV0Inner::NoValidate));
+
+        // Ethereum fee token contract address
+        let eth_fee_contract_address =
+            ContractAddress(PatriciaKey(StarkFelt::try_from(ETH_FEE_TOKEN_ADDRESS).unwrap()));
+
+        // Starknet fee token contract address
+        let strk_fee_contract_address =
+            ContractAddress(PatriciaKey(StarkFelt::try_from(STRK_FEE_TOKEN_ADDRESS).unwrap()));
+
+        let mut transaction = get_invoke_v3_dummy(Starknet::chain_id(), NONCE_ZERO);
+
+        if let starknet_api::transaction::InvokeTransaction::V3(tx) = &mut transaction.tx {
+            tx.sender_address = sender_address;
+        };
+
+        set_sender_erc20_balance_to_zero(sender_address, eth_fee_contract_address);
+
+        let initial_balance_vec = get_balance_contract_call(sender_address, strk_fee_contract_address);
+
+        // Ensure that strk fee token balance is not empty
+        pretty_assertions::assert_eq!(
+            initial_balance_vec,
+            vec![
+                Felt252Wrapper::from_hex_be("0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF").unwrap(),
+                Felt252Wrapper::from_hex_be("0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF").unwrap()
+            ]
+        );
+
+        assert_ok!(Starknet::invoke(none_origin, transaction));
+        let final_balance_vec = get_balance_contract_call(sender_address, strk_fee_contract_address);
+        assert!(final_balance_vec[1] == initial_balance_vec[1]);
+        assert!(final_balance_vec[0] < initial_balance_vec[0]);
+    });
+}
+
+#[test]
+fn given_hardcoded_contract_run_invoke_v3_without_strk_with_eth_fee_token_it_fails() {
```

Comment:
but my understranding was that currently on mainnet it is possible to pay with both ETH and STRK, why does it fails here?

Reply:
In my understanding (I check it on mainnet with ArgentX), when you pay fees with STRK, TXv3 is always used, and if you pay with ETH, it's a TXv1.
Also, in transaction field, you can't specify the token used as fees.

Reply:
> In my understanding (I check it on mainnet with ArgentX), when you pay fees with STRK, TXv3 is always used, and if you pay with ETH, it's a TXv1.

Hoooooo that's how it works!!

@apoorvsadana you were aware of this behavior?

Reply:
ya that was my doubt as well on the tg group, my next after this that when v1 is deprecated will STRK be on the only method of fee payment? or will we have proper paymasters by then?

Reply:
I think paymaster is at the applicative level, can it be part of the protocol? Of a broader "fee market"?

Reply:
I think paymaster will be in the protocol level but I haven't read a lot on this tbh. There's this post which shows the sequencer will need to be changed. -https://community.starknet.io/t/starknet-account-abstraction-model-part-2/839. And V3 txs also have paymaster data which isn't used now but might be used for this purpose eventually.